### ISSUE TITLE
Segment confusion (Same segment name different path)

### ISSUE THREAD

**Issue by davec21** [ionic-team/ionic#12831]
**Ionic version:**  (check one with "x")
[x] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report

**Current behavior:**
I am using deep linking and have come across an issue where two segments with matching page names causes the wrong page to load and updates the url to the wrong page.

**Expected behavior:**
Properly load the correct page.

**Steps to reproduce:**
1. Create two pages (page1 and page2)
2. Set the segment of page1 to { segment: 'index' }
3. Set the segment of page2 to { segment: 'training/index' }
4. Add a link from page1 to page2 (just to get to it)
5. Set page1 to be root
6. Run the app. The URL will be /#/index 
7. Click the link to get to page2 URL will update to /#/training/index
8. Refresh the page and notice page1 loads and the url updates to /#/index

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
cli packages: (C:\Source\deeplinkr\dashboard\node_modules)

    @ionic/cli-plugin-ionic-angular : 1.4.1
    @ionic/cli-utils                : 1.7.0
    ionic (Ionic CLI)               : 3.7.0

local packages:

    @ionic/app-scripts : 2.1.3
    Ionic Framework    : ionic-angular 3.6.0

System:

    Node : v6.11.0
    OS   : Windows 10
    npm  : 5.3.0
```

**Comment by kensodemann** [comment]
Hello! Thank you for opening an issue with us! Would you be able to provide a sample application via GitHub that demonstrates the issue you are having? Thanks for using Ionic!

**Comment by davec21** [comment]
https://github.com/davec21/dl-error

**Comment by kensodemann** [comment]
Thank you. I will have a look.

**Comment by astec** [comment]
Probably segment value should not allow `/` characters? As it complicates logic and could be a cause of confusions.

**Comment by davec21** [comment]
@astec how would you show sections/groupings? I am creating a system that has multiple similar named pages each in different 'sections'. For example:

/training/analytics
/user/analytics
/analytics

I don't want:
/traininganalytics
/useranalytics
/analytics

To me that looks messy as I use the groupings to keep things clean for me.

**Comment by astec** [comment]
I would simply use:
```
/training-analytics
/user-analytics
/analytics
```
because knowing how Ionic  breaks things this seems to me much more breaking-proof.

I have a picture of ideal URL routing in my head and I guess it's very close to yours but I afraid it's not going be in workable state in Ionic anytime soon.

For [DebtsTracker.io](https://debtstracker.io/) I decided to sacrifice beauty for reliability.

Once it's get stable (I still hope!) I'll improve URLs. For now for external links I link to server urls that do redirect to current schema for app urls (with #). So nice urls for emails, a bit ugly in the app.

**Comment by davec21** [comment]
For now that's what I am doing, but it doesn't make it any less of a bug.

**Comment by TreatDarren** [comment]
Yeah this is a real issue on desktop heavy apps with many layers, if you have an admin section of a normal site '/login' shouldn't be confused with. '/admin/login' and one should absolutely not have to do '/admin-login'

**Comment by danbucholtz** [comment]
Hi, this is a limitation of our current deep linking system. We're going to change how this all works in v4.

Right now, the deep linking system dynamically builds up a url by reading each chunk between `/` in the url from the right to left.

In the case in the OP, `index` is a segment, and then `training/index` is another segment. Since the url would be `/training/index`, it would match on `index` right away.

We're going to fix this in V4 but we cannot do it without an API change. The short term fix is to make your segment names totally unique. Don't use `nav` or `tabs` in them either, as that could have a host of other issues.

Thanks,
Dan
